planner/core/rule_column_pruning.go

```diff
@@ -424,6 +424,24 @@ func (p *LogicalWindow) extractUsedCols(parentUsedCols []*expression.Column) []*
 	return parentUsedCols
 }
 
+// PruneColumns implements LogicalPlan interface.
+func (p *LogicalLimit) PruneColumns(parentUsedCols []*expression.Column) error {
+	if len(parentUsedCols) == 0 { // happens when LIMIT appears in UPDATE.
+		return nil
+	}
+
+	p.schema = p.children[0].Schema().Clone()
```

Comment:
It's necessary to clone schema from its child, but I think it should not stay under `PruneColumns` because I would rather consider `PruneColumns` as an additional and optional optimization.

Could we do it in `logicalSchemaProducer. Schema` and `physicalSchemaProducer.Schema` as a default behavior?

```go
// Schema implements the Plan.Schema interface.
func (s *logicalSchemaProducer) Schema() *expression.Schema {
	if s.schema == nil {
		s.schema = s.baseLogicalPlan.children[0].Schema().Clone()
	}
	return s.schema
}

func (s *logicalSchemaProducer) OutputNames() types.NameSlice {
	if s.names == nil {
		s.names = s.baseLogicalPlan.children[0].OutputNames()
	}
	return s.names
}
```